CRADLE/CorrectBiasStored/calculateOneBP.pyx

```diff
@@ -56,30 +60,40 @@ cpdef performRegression(trainingSet, covariates, ctrlBWNames, ctrlScaler, experi
 
 	return COEFCTRL, COEFEXPR, ctrlPlotValues, experiPlotValues
 
-cpdef getReadCounts(bwFileName, trainingSet, scaler):
+def readCountData(bwFileName, trainingSet):
+	with pyBigWig.open(bwFileName) as bwFile:
+		if pyBigWig.numpy == 1:
+			for trainingRegion in trainingSet:
+				regionReadCounts = bwFile.values(trainingRegion.chromo, trainingRegion.analysisStart, trainingRegion.analysisEnd, numpy=True)
+				regionLength = trainingRegion.analysisEnd - trainingRegion.analysisStart
+				yield regionReadCounts, regionLength
+		else:
+			for trainingRegion in trainingSet:
+				regionReadCounts = np.array(
+					bwFile.values(trainingRegion.chromo, trainingRegion.analysisStart, trainingRegion.analysisEnd)
+				)
+				regionLength = trainingRegion.analysisEnd - trainingRegion.analysisStart
+				yield regionReadCounts, regionLength
+
+cpdef getReadCounts(rawReadCounts, rowCount, scaler):
 	cdef double [:] readCountsView
 	cdef int ptr
 	cdef int posIdx
 
-	readCounts = np.zeros(trainingSet.xRowCount, dtype=np.float64)
+	readCounts = np.zeros(rowCount, dtype=np.float64)
 	readCountsView = readCounts
```

Comment:
What is the use of `readCountsView `? It looks like some computation is performed to `readCountsView`, but it doesn't do anything with the returned value (readCounts)?

Reply:
readCounts and readCountsView are two interfaces to the same memory, so writing to one gives you the same result as writing to the other. But readCountsView is a [typed MemoryView](http://docs.cython.org/en/latest/src/userguide/memoryviews.html) which lets us access the underlying memory with lower overhead than by using the numpy array directly. This means that it's faster to write to the readCountsView, MemoryView, than to the readCounts numpy array. But in the end, readCounts has all the same numbers are readCountsView.

Reply:
Got you. Thanks!